Mathlib/RingTheory/Coalgebra.lean

```diff
@@ -0,0 +1,45 @@
+/-
+Copyright (c) 2023 Ali Ramsey. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Ali Ramsey
+-/
+import Mathlib.RingTheory.TensorProduct
+
+/-!
+# Definition and example of a coalgebra
+-/
+
+universe u v
+
+open scoped TensorProduct
+
+/-- A coalgebra over a commutative ring `R` is a module over `R` equipped with a coassociative
+comultiplication `Δ` and a counit `ε` obeying the left and right conunitality laws. -/
+class Coalgebra (R : Type u) (A : Type v) [CommRing R] [AddCommGroup A] [Module R A] where
+  /-- The comultiplication of the coalgebra -/
+  Δ : A →ₗ[R] A ⊗[R] A
+  /-- The counit of the coalgebra -/
+  ε : A →ₗ[R] R
+  /-- The comultiplication is coassociative -/
+  coassoc : TensorProduct.assoc R A A A ∘ₗ TensorProduct.map Δ .id ∘ₗ Δ =
```

Comment:
You should add some lemmas after this `class` that state the applied versions.

Reply:
Just to check: something like  
```lean
@[simp]
theorem coassoc_apply {R : Type u} {A : Type v} [CommRing R] [AddCommGroup A] [Module R A]
    [C : Coalgebra R A] (a : A) : (TensorProduct.assoc R A A A ∘ₗ TensorProduct.map C.Δ .id ∘ₗ C.Δ) a =
    (TensorProduct.map .id C.Δ ∘ₗ C.Δ) a := by
  exact LinearMap.congr_fun C.coassoc a
```
for each of the conditions in the definition?

Reply:
Not quite; the same proof, but the statement should be `∘ₗ`-free:
```lean
TensorProduct.assoc R A A A (TensorProduct.map C.Δ .id (C.Δ a)) = TensorProduct.map .id C.Δ (C.Δ a)
```

Reply:
Thank you! What's the etiquette on resolving conversations? I think I've added the lemmas correctly in the latest commit, but obviously I might be wrong and have made a mistake. Would I still resolve it and if I was wrong a new conversion would be opened?

Reply:
I don't know if we have a hard rule here; but I'd usually say it's fine for you to resolve things, especially if you add a comment saying "done" or similar at the same time